File: java/src/main/java/com/newnancity/newnanmanager/services/ServerService.java
package com.newnancity.newnanmanager.services;

import com.newnancity.newnanmanager.exceptions.NewNanManagerException;
import com.newnancity.newnanmanager.generated.api.ServerServiceApi;
import com.newnancity.newnanmanager.generated.model.CreateServerRequest;
import com.newnancity.newnanmanager.generated.model.ListServersResponse;
import com.newnancity.newnanmanager.generated.model.ServerDetailResponse;
import com.newnancity.newnanmanager.generated.model.ServerRegistry;
import com.newnancity.newnanmanager.generated.model.UpdateServerRequest;

/** Server registration and lookup service. */
public final class ServerService extends AbstractService {
    private final ServerServiceApi api;

    public ServerService(ServerServiceApi api) {
        this.api = api;
    }

    public ServerRegistry createServer(CreateServerRequest request) throws NewNanManagerException {
        return execute(() -> api.createServer(request).execute());
    }

    public ServerDetailResponse getServer(int id) throws NewNanManagerException {
        return getServer(id, null);
    }

    public ServerDetailResponse getServer(int id, Boolean detail) throws NewNanManagerException {
        return execute(() -> api.getServer(id).detail(detail).execute());
    }

    public ServerRegistry updateServer(int id, UpdateServerRequest request) throws NewNanManagerException {
        return execute(() -> api.updateServer(id, request).execute());
    }

    public void deleteServer(int id) throws NewNanManagerException {
        executeVoid(() -> api.deleteServer(id).execute());
    }

    public ListServersResponse listServers() throws NewNanManagerException {
        return listServers(null, null, null, null);
    }

    public ListServersResponse listServers(
            Integer page,
            Integer pageSize,
            String search,
            Boolean onlineOnly) throws NewNanManagerException {
        return execute(() -> api.listServers()
                .page(page)
                .pageSize(pageSize)
                .search(search)
                .onlineOnly(onlineOnly)
                .execute());
    }

    public ServerServiceApi raw() {
        return api;
    }
}

File: kotlin/src/main/kotlin/com/nanmanager/bukkit/services/ServerService.kt
package com.nanmanager.bukkit.services

import com.nanmanager.bukkit.http.HttpClient
import com.nanmanager.bukkit.models.*

/** 服务器管理服务；传输和模型解析由 OpenAPI 生成客户端负责。 */
class ServerService(httpClient: HttpClient) : BaseService(httpClient) {
    private val api get() = httpClient.generatedApis.servers

    fun createServer(request: CreateServerRequest): ServerRegistry {
        val body = toGenerated<com.newnanmanager.generated.models.CreateServerRequest>(request)
        return generated<ServerRegistry, com.newnanmanager.generated.models.ServerRegistry> {
            api.createServerWithHttpInfo(body)
        }
    }

    fun getServer(request: GetServerRequest): ServerDetailResponse {
        return generated<ServerDetailResponse, com.newnanmanager.generated.models.ServerDetailResponse> {
            api.getServerWithHttpInfo(request.id, request.detail)
        }
    }

    fun updateServer(request: UpdateServerRequest): ServerRegistry {
        val body = toGenerated<com.newnanmanager.generated.models.UpdateServerRequest>(request)
        return generated<ServerRegistry, com.newnanmanager.generated.models.ServerRegistry> {
            api.updateServerWithHttpInfo(request.id, body)
        }
    }

    fun deleteServer(request: DeleteServerRequest) {
        generatedVoid { api.deleteServerWithHttpInfo(request.id) }
    }

    fun listServers(request: ListServersRequest): ListServersResponse {
        return generated<ListServersResponse, com.newnanmanager.generated.models.ListServersResponse> {
            api.listServersWithHttpInfo(
                page = request.page,
                pageSize = request.pageSize,
                search = request.search,
                onlineOnly = request.onlineOnly
            )
        }
    }
}
